fix url not updating state

FlorianCassayre opened the pull request:
* fixes url history not updating the internal state
* closes inveniosoftware/react-invenio-app-ils#274

Tested on zenodo demo and on ILS. Also tested on CDS videos demo which has disabled support for url handling.

It may not be the best way to do it, but it works. I'm preventing the url from being updated twice: this is a workaround (I hope temporary) because the url is replaced twice, and even though the state is equivalent in both cases the order of the attributes is not.

Open to suggestions.

FlorianCassayre commented:
I used the following widespread design for requests:
* any request that is made triggers a loading state
* a completion (which can be successful or not) is accepted only if the corresponding request is the latest that was made, otherwise it is ignored

zzacharo commented on src/lib/state/reducers/results.js:
```diff
@@ -21,29 +21,38 @@ export default (state = {}, action) => {
-        },
-        error: action.payload,
-      };
+      if (state.queryState === action.payload.queryState) {
```
Maybe this can be done in the action level so that the reducer function stays pure :)

FlorianCassayre replied:
I'm not sure if it will work because we need to know the current state in order to update it accordingly.

ntarocco replied:
same as above

FlorianCassayre commented on src/lib/state/reducers/results.js:
```diff
@@ -21,29 +21,38 @@ export default (state = {}, action) => {
-        },
-        error: {},
-      };
+      if (state.queryState === action.payload.queryState) {
```
This is was I tried to explain in the above comment, namely there can be several concurrent requests running at the same time (for instance if the user goes back two pages, relatively quickly). In that case we only want to show trigger the completion event when the last request terminates, and ignore (or cancel) the others.
It's possible to use a counter for that purpose but a reference works too.

ntarocco replied:
I would suggest to make this explicit in the code, e.g.:
```
// comment explaining what is happening
isLatestRequest = state.queryState === action.payload.queryState
```

ntarocco reviewed (commented):
Thanks a lot for your explanation. If leave aside the race condition for a moment, I am not sure I understand what is causing the browser back issue.
It would be great if tomorrow we discuss this IRL! :)

FlorianCassayre commented:
PR was revised:
* the action `SET_QUERY_STATE` was _updating_ the state instead of _setting_ (= replacing) it, I fixed it and removed the newly created action
* the parameters of the url are now sorted lexicographically by their key, except for `q` which is meant to appear first
* removed the temporary parameter `replaceUrl` since now the replaced url has no change

ntarocco commented on src/lib/state/reducers/query.js:
```diff
@@ -92,7 +92,8 @@ export default (state = {}, action) => {
       return {
         ...state,
-        ...updateQueryState(state, action.payload, INITIAL_QUERY_STATE_KEYS),
+        ...INITIAL_QUERY_STATE,
```
As discussed, we should keep it to avoid that extra keys, coming externally (backend response), will be injected in the query state.

I would also rename the method `updateQueryState` to something more clear, e.g. `sanitize...`

FlorianCassayre commented:
Changes:
```diff
    case SET_QUERY_STATE:
      return {
        ...state,
        ...INITIAL_QUERY_STATE,
-       ...action.payload,
+       ...updateQueryState(INITIAL_QUERY_STATE, action.payload, INITIAL_QUERY_STATE_KEYS),
      };
```

`INITIAL_QUERY_STATE` is required for the filters to be correctly reset.
